### DISCUSSION TITLE
Git release tags and their appearance on the Git Graph

### DISCUSSION THREAD

**Discussion by plastikfan** [community/community#49512] [Pull Requests] [answered]
### Select Topic Area

Question

### Body

I've just started to use the PR model for my repos and am in the process of learning how to automate releases. I am experimenting on a test [repo](https://github.com/snivilised/lucretia) and have made a few 'fake' releases using github actions.

The state of the tree is now:

![lucretia-git-graph-v016](https://user-images.githubusercontent.com/49785914/223736373-a5a9bab5-80be-444a-82b6-419e8e73833c.jpg)

What I would like to know is, does this visualisation look correct for the Pull Request model, in particular those tagged commits (`49be359f`, `a09bd4cd` and `0dc838bf`).
__(please only consider the linear part as this is the point at which I activated branch protection rules)__

Those release commits were merged into master using the __Rebase and Merge__ option. The graph looks a bit weird, because it looks like commits are dangling references. I am assuming that because tags simply mark a point in the git history and releases are tagged on the relase branch, those dangling references are correct.

**Comment by airtower-luna** [comment, marked as answer]
The reason is right here:

> Those release commits were merged into master using the Rebase and Merge option.

Rebasing means that the commit on `master` is *not* the same commit as the tagged one. Rebasing creates a new commit with the same data (somewhat simplified). That's why you get the dangling: You have one "bump version" commit that's tagged, and *another one* that's on your `master` branch. As you want a linear history it's important to understand that this means your release tags point at commits *outside* that linear history.

The clean solution is usually to create the release tag *after* merging.

**Reply by plastikfan** [reply]
Ok thanks @airtower-luna. If I create the release tag after merging, then that means my release process should be:

- create a local release branch
- push that branch (now without a tag) (only the build ci process is triggered)
- once that is successful, rebase merge
- re pull the changes into local master branch
- tag release on (local) master, then push the tag -> this triggers the workflow that creates the changelog and the release

I had a feeling that my current workflow wasn't quite right (my spidey senses were tingling), and your suggestion corrects this. Please can you confirm that my proposed new release process will do the trick. Many thanks 🙏

**Reply by plastikfan** [reply]
mmm, I've just realised, that I don't necessarily need to have a release branch. I only need a release branch to make release specific changes. I typically update a version file, but if that is not required, then you can simply tag the local master branch, then push that, at which point the release is created. One for experimentation I think.

**Reply by airtower-luna** [reply]
Yep, both of those approaches will work. The second one (without a release branch) only if you don't have a branch protection rule that prevents direct pushes, though. :slightly_smiling_face:

**Reply by plastikfan** [reply]
> Yep, both of those approaches will work. The second one (without a release branch) only if you don't have a branch protection rule that prevents direct pushes, though. 🙂

Yeah I do have branch protection rules set up, but when I say push,  I only meant pushing tags, which I'm assuming is not subject to those protections rules. In my experimentation, I have been able to push tags to a protected branch.

**Reply by airtower-luna** [reply]
Tag protection rules are separate from branch protection rules. :wink: 

A tag is *never* on a branch. A tag points to a specific commit and that's it. If that commit is part (or HEAD) of a branch doesn't matter for the tag.

**Reply by plastikfan** [reply]
> Tag protection rules are separate from branch protection rules. 😉
> 
> A tag is _never_ on a branch. A tag points to a specific commit and that's it. If that commit is part (or HEAD) of a branch doesn't matter for the tag.

Thanks for your wisdom @airtower-luna, much appreciated.
